File: Symmetric Tree/SymmetricTree.py
class SymmetricTree(object):
    def isSymmetric(self, root):
        return self.isSame(root, root)
    
    def isSame(self,left,right):    
        if left is None and right is None:
            return True
        if left is not None and right is not None:
            return left.val == right.val and self.isSame(left.left, right.right) and self.isSame(left.right, right.left)
        return False
    
    def isSymmetricIteration(self, root):
        if root is None:
            return True
        queue = list()
        queue.append(root)
        queue.append(root)
        
        while len(queue)>0:
            left = queue.pop()
            right = queue.pop()
            if left == None and right == None:
                continue
            if left == None or right == None:
                return False
            if left.val == right.val:
                queue.append(left.left)
                queue.append(right.right)
                queue.append(left.right)
                queue.append(right.left)
            else:
                return False
        return True
    
    

File: Symmetric Tree/SymmetricTree.java
import java.util.LinkedList;

public class SymmetricTree {
    public boolean isSymmetric(TreeNode root) {
        return isSame(root, root);
    }
    
    private boolean isSame(TreeNode left, TreeNode right){
    	if(left == null && right == null) return true;
    	if(left != null && right != null){
    		return left.val == right.val && isSame(left.left, right.right) && isSame(left.right, right.left);
    	}
    	
    	return false;
    }
    
    public boolean isSymmetricIteration(TreeNode root){
    	if(root == null) return true;
    	LinkedList<TreeNode> queue = new LinkedList<TreeNode>();
    	queue.add(root);
    	queue.add(root);
    	
    	while(queue.size()> 0){
    		TreeNode left = queue.removeFirst();
    		TreeNode right = queue.removeFirst();
    		
    		if(left == null && right == null) continue;
    		if(left == null || right == null) return false;
    		
    		if(left.val == right.val){
    			queue.addLast(left.left);
    			queue.addLast(right.right);
    			queue.addLast(left.right);
    			queue.addLast(right.left);
    		}else{
    			return false;
    		}
    	}
    	
    	return true;    	
    }
}
